File: src/main/java/ca/exp/soundboard/rewrite/soundboard/SoundboardEntry.java
package ca.exp.soundboard.rewrite.soundboard;

import org.jnativehook.keyboard.NativeKeyEvent;

import java.io.File;
import java.io.UnsupportedEncodingException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Iterator;

public class SoundboardEntry {

    public int[] activationKeysNumbers;
    private String filePath;

    public SoundboardEntry(File file, int[] keys) {
        Path page = Paths.get(file.getAbsolutePath());
        filePath = page.toAbsolutePath().toString();

        activationKeysNumbers = keys;
        if (activationKeysNumbers == null) {
            activationKeysNumbers = new int[0];
        }
    }

    public boolean matchesPressed(ArrayList<Integer> pressedKeys) {
        int keysRemaining = activationKeysNumbers.length;

        if (keysRemaining == 0) {
            return false;
        }

        for (int actkey : activationKeysNumbers) {
            for (Iterator<Integer> localIterator = pressedKeys.iterator(); localIterator.hasNext();) {
                int presskey = localIterator.next();

                if (actkey == presskey) {
                    keysRemaining--;
                }
            }
        }

        return keysRemaining <= 0;
    }

    public int matchesHowManyPressed(ArrayList<Integer> pressedKeys) {
        int matches = 0;

        for (int i = 0; i < pressedKeys.size(); i++) {
            int key = pressedKeys.get(i);
            int keyCount = activationKeysNumbers.length;

            if (i < keyCount) {
                int hotkey = activationKeysNumbers[i];

                if (key == hotkey) {
                    matches++;
                }
            }
        }

        return matches;
    }

    public void play(AudioManager audioManager, boolean moddedSpeed) {
        File file = toFile();
        audioManager.playSoundClip(file, moddedSpeed);
    }

    public File toFile() {
        File file = new File(filePath);

        if (!file.exists()) {
            Path path = Paths.get(filePath);
            return path.toFile();
        }

        return file;
    }

    public void setFile(File file) {
        try {
            filePath = new String(file.getAbsolutePath().getBytes(Utils.fileEncoding));
        } catch (UnsupportedEncodingException exception) {
            exception.printStackTrace();
        }
    }

    public String getFilePath() {
        return filePath;
    }

    public String getFileName() {
        char separator = File.separatorChar;
        return filePath.substring(filePath.lastIndexOf(separator) + 1);
    }

    public int[] getActivationKeys() {
        return activationKeysNumbers;
    }

    public void setActivationKeys(int[] activationKeys) {
        activationKeysNumbers = activationKeys;
    }

    public String getActivationKeysAsReadableString() {
        String activationKey = "";

        if (activationKeysNumbers.length == 0) {
            return activationKey;
        }

        for (int integer : getActivationKeys()) {
            activationKey = activationKey.concat(NativeKeyEvent.getKeyText(integer) + "+");
        }

        activationKey = activationKey.substring(0, activationKey.length() - 1);
        return activationKey;
    }
}

File: src/main/java/ca/exp/soundboard/soundboard/SoundboardEntry.kt
package ca.exp.soundboard.soundboard

import org.jnativehook.keyboard.NativeKeyEvent
import java.io.File

class SoundboardEntry(var file: File, keys: IntArray?) {
    var activationKeys = keys ?: IntArray(0)

    fun matchesPressed(pressedKeys: ArrayList<Int>): Boolean {
        var keysRemaining = activationKeys.size
        if (keysRemaining == 0) return false

        for (key in activationKeys) {
            val localIterator = pressedKeys.iterator()

            while (localIterator.hasNext()) {
                val pressedKey = localIterator.next()

                if (key == pressedKey) {
                    keysRemaining--
                }
            }
        }

        return keysRemaining <= 0
    }

    fun getMatchCount(pressedKeys: ArrayList<Int>): Int {
        var matches = 0

        for (i in pressedKeys.indices) {
            val key = pressedKeys[i]
            val keyCount = activationKeys.size

            if (i < keyCount) {
                val hotkey = activationKeys[i]

                if (key == hotkey) {
                    matches++
                }
            }
        }

        return matches
    }

    fun play(audioManager: AudioManager, moddedSpeed: Boolean) {
        audioManager.playSoundClip(file, moddedSpeed)
    }

    val activationKeysString: String
        get() {
            var activationKeysString = ""

            if (activationKeys.isEmpty()) {
                return activationKeysString
            }

            for (keyCode in activationKeys) {
                activationKeysString = activationKeysString + NativeKeyEvent.getKeyText(keyCode) + "+"
            }

            activationKeysString = activationKeysString.substring(0, activationKeysString.length - 1)
            return activationKeysString
        }
}
